php_phongo.c

```diff
@@ -1777,140 +1778,146 @@ static bool php_phongo_apply_rp_options_to_uri(mongoc_uri_t* uri, bson_t* option
-			mongoc_read_prefs_set_mode(new_rp, MONGOC_READ_SECONDARY_PREFERRED);
-		} else if (0 == strcasecmp("nearest", str)) {
-			mongoc_read_prefs_set_mode(new_rp, MONGOC_READ_NEAREST);
-		} else {
-			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Unsupported %s value: '%s'", bson_iter_key(&iter), str);
-			mongoc_read_prefs_destroy(new_rp);
+			str = bson_iter_utf8(&iter, NULL);
+
+			if (0 == strcasecmp("primary", str)) {
+				mongoc_read_prefs_set_mode(new_rp, MONGOC_READ_PRIMARY);
+			} else if (0 == strcasecmp("primarypreferred", str)) {
+				mongoc_read_prefs_set_mode(new_rp, MONGOC_READ_PRIMARY_PREFERRED);
+			} else if (0 == strcasecmp("secondary", str)) {
+				mongoc_read_prefs_set_mode(new_rp, MONGOC_READ_SECONDARY);
+			} else if (0 == strcasecmp("secondarypreferred", str)) {
+				mongoc_read_prefs_set_mode(new_rp, MONGOC_READ_SECONDARY_PREFERRED);
+			} else if (0 == strcasecmp("nearest", str)) {
+				mongoc_read_prefs_set_mode(new_rp, MONGOC_READ_NEAREST);
+			} else {
+				phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Unsupported %s value: '%s'", bson_iter_key(&iter), str);
+				mongoc_read_prefs_destroy(new_rp);
 
-			return false;
+				return false;
+			}
+
+			ignore_slaveok = true;
 		}
-	}
 
-	if (bson_iter_init_find_case(&iter, options, MONGOC_URI_READPREFERENCETAGS)) {
-		bson_t         tags;
-		uint32_t       len;
-		const uint8_t* data;
+		if (!strcasecmp(key, MONGOC_URI_READPREFERENCETAGS)) {
+			bson_t         tags;
+			uint32_t       len;
+			const uint8_t* data;
 
-		if (!BSON_ITER_HOLDS_ARRAY(&iter)) {
-			PHONGO_URI_INVALID_TYPE(iter, "array");
-			mongoc_read_prefs_destroy(new_rp);
+			if (!BSON_ITER_HOLDS_ARRAY(&iter)) {
+				PHONGO_URI_INVALID_TYPE(iter, "array");
+				mongoc_read_prefs_destroy(new_rp);
 
-			return false;
-		}
+				return false;
+			}
 
-		bson_iter_array(&iter, &len, &data);
+			bson_iter_array(&iter, &len, &data);
 
-		if (!bson_init_static(&tags, data, len)) {
-			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Could not initialize BSON structure for read preference tags");
-			mongoc_read_prefs_destroy(new_rp);
+			if (!bson_init_static(&tags, data, len)) {
+				phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Could not initialize BSON structure for read preference tags");
+				mongoc_read_prefs_destroy(new_rp);
 
-			return false;
-		}
+				return false;
+			}
 
-		if (!php_phongo_read_preference_tags_are_valid(&tags)) {
-			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Read preference tags must be an array of zero or more documents");
-			mongoc_read_prefs_destroy(new_rp);
+			if (!php_phongo_read_preference_tags_are_valid(&tags)) {
+				phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Read preference tags must be an array of zero or more documents");
+				mongoc_read_prefs_destroy(new_rp);
 
-			return false;
+				return false;
+			}
+
+			mongoc_read_prefs_set_tags(new_rp, &tags);
```

Comment:
I've confirmed that it's safe to call `mongoc_read_prefs_set_tags` twice, on the off chance there are duplicate `readPreferenceTags` options.

Reply:
That didn't occur to me, thanks for confirming. I'll add a test that overwrites `readPreferenceTags` as you suggested above, which will continuously ensure this works as expected.

Reply:
It would surely be a bug in libmongoc if this leaked, but I just thought to double-check since it's a non-scalar option, unlike most of the other things we're setting in this PR (e.g. mode strings and integers).